In [6]:
!pip3 install --user okpy

from lecture import *

You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


# Introduction to programming in Python

# Lecture 1

## Learning objectives:

* You will understand that Python will help you defy gravity.
* You will know how to execute Python statements from within Jupyter.
* Learn what a program variable is and how to express a mathematical expression in code.
* Print program outputs.
* Access mathematical functions from a Python module.
* Be able to write your own *function*.
* Know how to form a *condition* using a *boolean expression*.
* Be able to use a conditional expression in combination with a *while-loop* to perform repetitive tasks.
* Be able to store data elements within a Python *list*.
* Be able to use a *for-loop* to iterate, and perform some task, over a *list* of elements.

# Introduction to programming in Python
### [Gerard Gorman](http://www.imperial.ac.uk/people/g.gorman)

# Lecture 2: Conditional expressions, loops and lists

```python
import antigravity
```

![import antigravity](https://imgs.xkcd.com/comics/python.png)

## Programming a mathematical formula

Here is a formula for the position of a ball in vertical motion, starting at ground level (i.e. $y=0$) at time $t=0$:
     $$ y(t) = v_0t- \frac{1}{2}gt^2 $$

where:

* $y$ is the height (position) as a function of time $t$
* $v_0$ is the initial velocity (at $t=0$)
* $g$ is the acceleration due to gravity

The computational task is: given $v_0$, $g$ and $t$, compute the value $y$. 

**How do we program this task?** A program is a sequence of instructions given to the computer. However, while a programming language is much **simpler** than a natural language, it is more **pedantic**. Programs must have correct syntax, i.e., correct use of the computer language grammar rules, and no misprints.

So let's execute a Python statement based on this example. Evaluate $y(t) = v_0t- \frac{1}{2}gt^2$ for $v_0=5$, $g=9.81$ and $t=0.6$. If you were doing this on paper you would probably write something like this: $$ y = 5\cdot 0.6 - {1\over2}\cdot 9.81 \cdot 0.6^2.$$ Happily, writing this in Python is very similar:

In [7]:
# Comment: This is a 'code' cell within Jupyter notebook.
# Press shift-enter to execute the code within this kind of
# cell, or click on the 'Run' widget on the Jupyter tool bar above.

print(5*0.6 - 0.5*9.81*0.6**2)

1.2342


## Exercise 1.1: Open a code cell and write some code.
* Navigate the [Jupyter](http://jupyter.org/) tool bar to "Insert"->"Insert Cell Below". Note from the tool bar that you can select a cell to be one of 'Code' (this is the default), 'Markdown' (this cell is written in [markdown](https://en.wikipedia.org/wiki/Markdown) - double click this cell to investigate further), 'Raw NBConvert' or 'Heading' (decrepit).
* Cut&paste the code from the previous cell into your newly created code cell below. Make sure it runs!
* To see how important it is to use the correct [syntax](https://en.wikipedia.org/wiki/Syntax), replace `**` with `^` in your code and try running the cell again. You should see something like the following:

```python
>>> print(5*0.6 - 0.5*9.81*0.6^2)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-40e93484ac5e> in <module>()
----> 1 print(5*0.6 - 0.5*9.81*0.6^2)

TypeError: unsupported operand type(s) for ^: 'float' and 'int'
```
* Undo that change so your code is working again; now change 'print' to 'write' and see what happens when you run the cell. You should see something like:

```python
>>> write(5*0.6 - 0.5*9.81*0.6**2)

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-5-492c3eff3ad9> in <module>()
----> 1 write(5*0.6 - 0.5*9.81*0.6**2)

NameError: name 'write' is not defined
```

While a human might still understand these statements, they do not mean anything to the Python interpreter. Rather than throwing your hands up in the air whenever you get an error message like the above (you are going to see many during the course of these lectures!!!) train yourself to read error messages carefully to get an idea what it is complaining about and re-read your code from the perspective of the Python interpreter.

Error messages can look bewildering and even frustrating at first, but it gets much **easier with practise**.

In [8]:
print(5*0.6 - 0.5*9.81*0.6**2)

1.2342


In [9]:
print(5*0.6 - 0.5*9.81*0.6^2)

TypeError: unsupported operand type(s) for ^: 'float' and 'int'

In [11]:
print("hello world")

hello world


## Storing numbers in variables
From mathematics you are already familiar with variables (e.g. $v_0=5,\quad g=9.81,\quad t=0.6,\quad y = v_0t -{1\over2}gt^2$) and you already know how important they are for working out complicated problems. Similarly, you can use variables in a program to make it easier to read and understand.

In [10]:
v0 = 5
g = 9.81
t = 0.6
y = v0*t - 0.5*g*t**2
print(y)

1.2342


This program spans several lines of text and uses variables, otherwise the program performs the same calculations and gives the same output as the previous program.

In mathematics we usually use one letter for a variable, resorting to using the Greek alphabet and other characters for more clarity. The main reason for this is to avoid becoming exhausted from writing when working out long expressions or derivations. However, when programming you should use more descriptive names for variable names. This might not seem like an important consideration for the trivial example here but it becomes increasingly important as the program gets more complicated and if someone else has to read your code.

### Good variable names make a program easier to understand!

Permitted variable names include:

* One-letter symbols.
* Words or abbreviation of words.
* Variable names can contain a-z, A-Z, underscore ("'_'") and digits 0-9, **but** the name cannot start with a digit.

Variable names are case-sensitive (i.e. "'a'" is different from "'A'"). Let's rewrite the previous example using more descriptive variable names:

In [12]:
initial_velocity = 5
acceleration_of_gravity = 9.81
TIME = 0.6
VerticalPositionOfBall = initial_velocity*TIME - 0.5*acceleration_of_gravity*TIME**2
print(VerticalPositionOfBall)

1.2342


Certain words have a **special meaning** in Python and **cannot be used as variable names**. These are: *and, as, assert, break, class, continue, def, del, elif, else, except, exec, finally, for, from, global, if, import, in, is, lambda, not, or, pass, print, raise, return, try, with, while,* and *yield*.

## Adding comments to code

Not everything written in a computer program is intended for execution. In Python anything on a line after the '#' character is ignored and is known as a **comment**. You can write whatever you want in a comment. Comments are intended to be used to explain what a snippet of code is intended for. It might for example explain the objective or provide a reference to the data or algorithm used. This is both useful for you when you have to understand your code at some later stage, and indeed for whoever has to read and understand your code later.

In [13]:
# Program for computing the height of a ball in vertical motion.
v0 = 5    # Set initial velocity in m/s.
g = 9.81  # Set acceleration due to gravity in m/s^2.
t = 0.6   # Time at which we want to know the height of the ball in seconds.
y = v0*t - 0.5*g*t**2 # Calculate the vertical position
print(y)

1.2342


## Exercise 1.2: Convert from meters to British length units
Here in the UK we are famous for our love of performing mental arithmetic. That is why we still use both imperial and metric measurement systems - hours of fun entertainment for the family switching back and forth between the two.

Make a program where you set a length given in meters and then compute and write out the corresponding length measured in:
* inches (one inch is 2.54 cm)
* feet (one foot  is 12 inches)
* yards (one foot is 12 inches, one yard is 3 feet)
* miles (one British mile is 1760 yards)

Note: In this course we are using [okpy](https://okpy.org/) to automated assessment scoring. Therefore, while it is important to always carefully follow the instructions of a question, it is particularly important here so that okpy can recognize the validity of your answer. The conversion to inches are done for you to illustrate what is required.

In [14]:
meters = 640

# 1 inch = 2.54 cm. Remember to convert from 2.54 cm to 0.0254 m here.
inches = meters/0.0254

# Uncomment and complete the following code.
feet = inches/12

yards = feet/3

miles = yards/1760

In [15]:
grade = ok.grade('question-1_2')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



## Formatted printing style
Often we want to print out results using a combination of text and numbers, e.g. "'At t=0.6 s, y is 1.23 m'". Particularly when printing out floating point numbers we should **never** quote numbers to a higher accuracy than they were measured. Python provides a *printf formatting* syntax exactly for this purpose. We can see in the following example that the *slot* `%g` was used to express the floating point number with the minimum number of significant figures, and the *slot* `%.2f` specified that only two digits are printed out after the decimal point.

In [16]:
print("At t=%g s, y is %.2f m." % (t, y))

At t=0.6 s, y is 1.23 m.


Notice in this example how the values in the tuple `(t, y)` are inserted into the *slots*.

Sometimes we want a multi-line output. This is achieved using a triple quotation (*i.e.* `"""`):

In [17]:
print("""At t=%f s, a ball with
initial velocity v0=%.3E m/s
is located at the height %.2f m.
""" % (t, v0, y))

At t=0.600000 s, a ball with
initial velocity v0=5.000E+00 m/s
is located at the height 1.23 m.



## Exercise 1.3: Compute the air resistance on a football
The drag force, due to air resistance, on an object can be expressed as
$$F_d = \frac{1}{2}C_D\rho AV^2$$
where:
* $\rho$ is the density of the air,
* $V$ is the velocity of the object,
* $A$ is the cross-sectional area (normal to the velocity direction),
* and $C_D$ is the drag coefficient, which depends on the shape of the object and the roughness of the surface.

Complete the following code that computes the drag force. 

In [18]:
# Football example

# import pi from Python's math library
from math import pi

density = 1.2      # units of kg m^{−3}$
ball_radius = 0.11 # m
A = pi*ball_radius**2 # Cross sectional area of a sphere
mass = 0.43        # kg
C_D = 0.2          # Drag coefficient

V = 50.8           # m/s (fastest recorded speed of football)

# Uncomment and complete the following code.
F_d = 0.5* density * C_D * A * V**2 #Drage Force
print("The drag force due to air resistance is : %.1f Newtons" %(F_d))

# Challenge yourself to use the formatted print statement
# shown above to write out the forces with one decimal in
# units of Newton ($N = kgm/s^2$).

The drag force due to air resistance is : 11.8 Newtons


In [19]:
grade = ok.grade('question-1_3')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



## How are arithmetic expressions evaluated?
Consider the random mathematical expression, ${5\over9} + 2a^4/2$, implemented in Python as `5.0/9 + 2*a**4/2`.

The rules for evaluating the expression are the same as in mathematics: proceed term by term (additions/subtractions) from the left, compute powers first, then multiplication and division. Therefore in this example the order of evaluation will be:

1. `r1 = 5.0/9`
2. `r2 = a**4`
3. `r3 = 2*r2`
4. `r4 = r3/2`
5. `r5 = r1 + r4`

Use parenthesis to override these default rules. Indeed, many programmers use parenthesis for greater clarity.

## Exercise 1.4: Compute the growth of money in a bank
Let *p* be a bank's interest rate in percent per year. An initial amount $A_0$ has then grown to $$A_n = A_0\left(1+\frac{p}{100}\right)^n$$ after *n* years. Write a program for computing how much money 1000 euros have grown to after three years with a 5% interest rate.

In [20]:
# Complete the code commented out below (don't change variable names!)

p = 5 #interest rate
A_0 = 1000 #initial amount
n = 3 #no of years
A_n = A_0*(1+(p/100))**n

print("The amount of money in the account after %d years is: %.2f euros" % (n, A_n))

The amount of money in the account after 3 years is: 1157.63 euros


In [21]:
grade = ok.grade('question-1_4')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



## Standard mathematical functions
What if we need to compute $\sin x$, $\cos x$, $\ln x$, etc. in a program? Such functions are available in Python's *math module*. In fact there is a vast universe of functionality for Python available in modules. We just *import* in whatever we need for the task at hand.

In this example we compute $\sqrt{2}$ using the *sqrt* function in the *math* module:

In [22]:
import math
r = math.sqrt(2)
print(r)

1.4142135623730951


or:

In [23]:
from math import sqrt
r = sqrt(2)
print(r)

1.4142135623730951


or:

In [24]:
from math import *   # import everything in math
r = sqrt(2)
print(r)

1.4142135623730951


Another example:

In [25]:
from math import sin, cos, log
x = 1.2
print(sin(x)*cos(x) + 4*log(x))   # log is ln (base e)

1.0670178174513938


## Exercise 1.5: Evaluate a Gaussian function

The bell-shaped Gaussian function,
$$f(x)=\frac{1}{\sqrt{2\pi}s}\exp\left(-\frac{1}{2} \left(\frac{x-m}{s}\right)^2\right)$$
is one of the most widely used functions in science and technology. The parameters $m$ and $s$ are real numbers, where $s$ must be greater than zero. Write a program for evaluating this function when $m = 0$, $s = 2$, and $x = 1$. Verify the program's result by using a calculator.

In [26]:
# Uncomment and complete the code below (don't change variable names!)

from math import pi, exp, sqrt

m=0
s=2
x=1

f_x = 1/(sqrt(2*pi)*s) * exp(-0.5*((x-m)/s)**2)
print("%.3f" % (f_x))

0.176


In [27]:
grade = ok.grade('question-1_5')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



## Exercise 1.6: Find errors in the coding of a formula
Given a quadratic equation,
$$ax^2 + bx + c = 0,$$
$$x1 = \frac{−b + \sqrt{b^2 −4ac}}{2a},$$ and
$$x2 = \frac{−b − \sqrt{b^2 −4ac}}{2a}.$$

Uncomment and fix the errors in the following code.

In [28]:
a = 2; b = 1; c = -2
from math import sqrt
q = sqrt(b*b - 4*a*c)
x1 = (-b + q)/(2*a)
x2 = (-b - q)/(2*a)
print("%.3f,%.3f" %(x1, x2))

0.781,-1.281


In [29]:
grade = ok.grade('question-1_6')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



## Functions

We have already used Python functions, e.g. `sqrt` from the `math` module above. In general, a function is a collection of statements we can execute wherever and whenever we want. For example, consider any of the formula you implemented above. 

Functions can take any number of inputs (called *arguments*) to produce outputs. Functions help to organize programs, make them more understandable, shorter, and easier to extend. Wouldn't it be nice to implement it just once and then be able to use it again any time you need it, rather than having to write out the whole formula again?

For our first example we will reuse the formula for the position of a ball in vertical motion, which we've seen at the beginning of the document. 

In [30]:
# Function to compute height of ball.
def ball_height(v0, t, g=9.81):
    """Function to calculate height of ball.
    
    Parameters
    ----------
    v0 : float
        Set initial velocity (units, m/s).
    t : float
        Time at which we want to know the height of the ball (units, seconds).
    g : float, optimal
        Acceleration due to gravity, by default 9.81 m/s^2.

    Returns
    -------
    float
        Height of ball in meters.
    """

    height = v0*t - 0.5*g*t**2
    
    return height

Lets break this example down:
* Function header:
 * Functions start with *def* followed by the name you want to give the function (ball_height in this case).
 * Following the name, you have `(...):` containing some number of function `arguments`.
 * In this case `v0` and `t` are *position arguments* while `g` is known as a *keyword argument* (more about this later).
* Function body.
 * The first thing to notice is that the body of the function is indented one level.
 * Best practice is to include a [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html) to explain to others (or remind your future self) how the function should be used.
 * The function output is passed back via the `return` statement
 
Notice that this just defines the function. Nothing is actually executed until you actually *use* the function:

In [31]:
print("Ball height: %g meters."%ball_height(5, 0.6))

Ball height: 1.2342 meters.


No return value implies that `None` is returned. `None` is a special Python object that represents an ”empty” or undefined value. It is surprisingly useful and we will use it a lot later.

Functions can also return multiple values. Let's extend the previous example to calculate the ball velocity as well as the height:

In [32]:
# Function to compute height of ball.
def ball_height_velocity(v0, t, g=9.81):
    """Function to calculate height and velocity of ball.
    
    Parameters
    ----------
    v0 : float
        Set initial velocity (units, m/s).
    t : float
        Time at which we want to know the height of the ball (units, seconds).
    g : float, optimal
        Acceleration due to gravity, by default 9.81 m/s^2.

    Returns
    -------
    float
        Height of ball in meters.
    float
        Velocity of ball in m/s.
    """

    height = v0*t - 0.5*g*t**2
    velocity = v0 - g*t
    
    return height, velocity

h, v = ball_height_velocity(5, 0.6)

print("Ball height: %g meters."%h)
print("Ball velocity: %g m/s."%v)

Ball height: 1.2342 meters.
Ball velocity: -0.886 m/s.


## Scope: Local and global variables

Variables defined within a function are said to have *local scope*. That is to say that they can only be referenced within that function. Consider the example function defined above where we used the *local* variable *height*. You can see that if you try to print the variable height outside the function you will get an error.

```python
print(height)

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-8-aa6406a13920> in <module>()
----> 1 print(height)

NameError: name 'height' is not defined
```

## Keyword arguments and default input values

Functions can have arguments of the form variable_name=value and are called keyword arguments:

In [33]:
def somefunc(arg1, arg2, kwarg1=True, kwarg2=0):
    print(arg1, arg2, kwarg1, kwarg2)

somefunc("Hello", [1,2])   # Note that we have not specified inputs for kwarg1 and kwarg2

Hello [1, 2] True 0


In [34]:
somefunc("Hello", [1,2], kwarg1="Hi")

Hello [1, 2] Hi 0


In [35]:
somefunc("Hello", [1,2], kwarg2="Hi")

Hello [1, 2] True Hi


In [36]:
somefunc("Hello", [1,2], kwarg2="Hi", kwarg1=6)

Hello [1, 2] 6 Hi


If we use variable_name=value for all arguments, their sequence in the function header can be in any order.

In [37]:
somefunc(kwarg2="Hello", arg1="Hi", kwarg1=6, arg2=[2])

Hi [2] 6 Hello


## Exercise 1.7: Implement a Gaussian function

Create a Python function to compute the Gaussian: 
$$f(x)=\frac{1}{s\sqrt{2\pi}}\exp\left(-\frac{1}{2} \left(\frac{x-m}{s}\right)^2\right)$$

In [38]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def gaussian(x, m=0, s=1):
    from math import pi, sqrt, exp
    a = exp(-0.5*((x-m)/s)**2)
    b = 1/(s*sqrt(2*pi))
    f_x = a*b
    return f_x
f_x = gaussian(1)
print ("%.3f" %(f_x))

0.242


In [39]:
ok.grade('question-1_7')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 11
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 11}

## Exercise 1.8: How to cook the perfect egg

You just started University and moved away from home. You're trying to impress your new flatmates by cooking brunch. Write a python script to help you cook the perfect eggs! 


You know from A-levels that when the temperature exceeds a critical point, the proteins in the egg first denature and then coagulate, and the process becomes faster as the temperature increases. In the egg white, the proteins start to coagulate for temperatures above 63$^\circ$C, while in the yolk the proteins start to coagulate for temperatures above 70$^\circ$C. 

The time `t` (in seconds) it takes for the centre of the yolk to reach the temperature `Ty` (in degrees Celsius) can be expressed as: 

$$t = \frac{M^{2/3}c \rho^{1/3}}{K \pi^2 (4 \pi /3)^{2/3} } ln \left[0.76 \frac{T_0-T_w}{T_y-T_w}\right]$$

where:
* $M$ is the mass of the egg;
* $\rho$ is the density;
* $c$ is the specific heat capacity;
* $K$ is thermal conductivity.
* $Tw$ temperature of the boiling water (in C degrees) 
* $T0$ is the initial temeprature of the egg (in C degrees), before being put in the water.

Write a function that returns the time `t` needed for the egg to cook, knowing that `Tw` = 100$^\circ$ C, `M` = 50 g, `rho` = 1.038 gcm$^{−3}$, `c` = 3.7 Jg$^{−1}$K$^{−1}$, and `K` = 5.4 · 10$^{−3}$Wcm$^{−1}$K$^{−1}$. Find `t` for an egg taken from the fridge (`T0` = 4 C) and for one at room temperature (`T0` = 20 C). 

Hint: You do not need to do any unit conversion. `Ty` = 70 $^\circ$ C. for a perfect soft boiled egg. 

In [40]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

from math import pi, log

def perfect_egg(T0, M=5, Tw=100, rho=1.038, c=3.7, K=5.4E-3, Ty=70):
    m = (M**(2/3)*c*rho**(1/3))/(K*pi**2*(4*pi/3)**(2/3))
    q = log(0.76*((T0-Tw)/(Ty-Tw)))
    t = m*q
    return t
print("time for an egg taken into the fridge at 4C is: %.2f" %(perfect_egg(4)))
print("time for an egg taken into the fridge at 20C is: %.2f" %(perfect_egg(20)))

time for an egg taken into the fridge at 4C is: 70.29
time for an egg taken into the fridge at 20C is: 55.87


In [41]:
ok.grade('question-1_8')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 1}

## Exercise 1.9: Kepler's third law

You were selected to be the next astronaut to go to Mars. Congratulations! 

Kepler's third law expresses the relationship between the distance of planets from the Sun, $a$, and their orbital periods, $P$

$$ P^2 = \frac{4\pi^2}{G(m_1+m_2)}a^3  $$

where
* P is the period (in seconds)
* G is the gravitational constant ( = 6.67 × 10$^{-11}$ m$^3$kg$^{-1}$s$^{-2}$)
* m$_1$ is the mass of the planet (in kg)
* m$_2$ is the mass of the Sun ( = 2x10$^{30}$ kg)
* a is the distance between the planet and the Sun (in m)

How many Earth birthdays will you celebrate during your 10 Mars years mission? Write a python code that will calculate the the period on Earth, `P_earth`, the period on Mars, `P_mars`, and how many Earth years are equivalent to 10 years on Mars, `birthdays`.

You know that:
* The average distance between the Earth and the Sun is $a$ = 1.5x10$^{11}$m;
* The average distance between Mars and the Sun is 0.5 larger than the Eart-Sun distance;
* The mass of the Earth is m$_1$ = 6x10$^{24}$;
* Mars's mass is about 10% of the Earth's mass.

Hint: You do not need to do any unit conversion. 

In [42]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

from math import pi, sqrt

def period(a, G=6.67*10**-11, m1=6*10**24 , m2=2*10**30):
    P_earth = sqrt((4*pi**2*a**3)/(G*(m1+m2))) 
    P_mars = sqrt((4*pi**2*(a*1.5)**3)/(G*(0.1*m1+m2)))
    birthdays = (P_mars*10/P_earth)
    return P_mars, P_earth, birthdays
a, b, c = period(1.5*10**11)
print("""The period on Mars is %.3f
The period on Earth is %.3f
The number of birthdays is %d""" %(a,b,c))

The period on Mars is 58059817.395
The period on Earth is 31603718.930
The number of birthdays is 18


In [43]:
ok.grade('question-1_9')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 1}

## Boolean expressions
An expression with value *true* or *false* is called a boolean expression. Example expressions for what you would write mathematically as
$C=40$, $C\ne40$, $C\ge40$, $C\gt40$ and $C\lt40$ are:

```python
C == 40 # Note: the double == checks for equality!
C != 40 # This could also be written as 'not C == 4'
C >= 40
C > 40
C < 40
```

We can test boolean expressions in a Python shell:

In [44]:
C = 41
print("C != 40: ", C != 40)
print("C < 40: ", C < 40)
print("C == 41: ", C == 41)

C != 40:  True
C < 40:  False
C == 41:  True


Several conditions can be combined with the special 'and' and 'or' keywords into a single boolean expression:

* Rule 1: (**C1** *and* **C2**) is *True* only if both **C1** and **C2** are *True*
* Rule 2: (**C1** *or* **C2**) is *True* if either **C1** or **C2** are *True*

Examples:

In [45]:
x=0; y=1.2
print (x >= 0 and y < 1)

False


## Exercise 1.10: Values of boolean expressions
Add a comment to the code below to explain the outcome of each of the boolean expressions:

In [46]:
C = 41

print("Case 1: ", C == 40)
print("Case 2: ", C != 40 and C < 41)
print("Case 3: ", C != 40 or C < 41)
print("Case 4: ", not C == 40)
print("Case 5: ", not C > 40)
print("Case 6: ", C <= 41)
print("Case 7: ", not False)
print("Case 8: ", True and False)
print("Case 9: ", False or True)
print("Case 10: ", False or False or False)
print("Case 11: ", True and True and False)
print("Case 12: ", False == 0)
print("Case 13: ", True == 0)
print("Case 14: ", True == 1)

Case 1:  False
Case 2:  False
Case 3:  True
Case 4:  True
Case 5:  False
Case 6:  True
Case 7:  True
Case 8:  False
Case 9:  True
Case 10:  False
Case 11:  False
Case 12:  True
Case 13:  False
Case 14:  True


## Loops
Suppose we want to make a table of Celsius and Fahrenheit degrees:
```
             -20  -4.0
             -15   5.0
             -10  14.0
              -5  23.0
               0  32.0
               5  41.0
              10  50.0
              15  59.0
              20  68.0
              25  77.0
              30  86.0
              35  95.0
              40 104.0
              ```

How do we write a program that prints out such a table?

We know from the last lecture how to make one line in this table:

In [47]:
C = -20
F = 9.0/5*C + 32
print(C, F)

-20 -4.0


We can just repeat these statements:

In [48]:
C=-20; F=9.0/5*C+32; print(C,F)
C=-15; F=9.0/5*C+32; print(C,F)
C=-10; F=9.0/5*C+32; print(C,F)
C=-5; F=9.0/5*C+32; print(C,F)
C=0; F=9.0/5*C+32; print(C,F)
C=5; F=9.0/5*C+32; print(C,F)
C=10; F=9.0/5*C+32; print(C,F)
C=15; F=9.0/5*C+32; print(C,F)
C=20; F=9.0/5*C+32; print(C,F)
C=25; F=9.0/5*C+32; print(C,F)
C=30; F=9.0/5*C+32; print(C,F)
C=35; F=9.0/5*C+32; print(C,F)
C=40; F=9.0/5*C+32; print(C,F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


So we can see that works but its **very boring** to write and very easy to introduce a misprint.

**You really should not be doing boring repetitive tasks like this.** Spend your time instead looking for a smarter solution. When programming becomes boring, there is usually a construct that automates the writing. Computers are very good at performing repetitive tasks. For this purpose we use **loops**.

## The while loop (and the significance of indentation)
A while loop executes repeatedly a set of statements as long as a **boolean** (i.e. *True* / *False*) condition is *True*

```
while condition:
    <statement 1>
    <statement 2>
    ...
<first statement after loop>
```

Note that all statements to be executed within the loop must be indented by the same amount! The loop ends when an unindented statement is encountered.

At this point it is worth noticing that **blank spaces may or may not be important** in Python programs. These statements are equivalent (blanks do not matter):

In [49]:
v0=3
v0  =  3
v0=   3
# The computer does not care but this formatting style is
# considered clearest for the human reader.
v0 = 3

Here is a while loop example where blank spaces really do matter:

In [50]:
counter = 0
while counter <= 10:
    counter = counter + 1
print(counter)

11


Let's take a look at what happens when we forget to indent correctly:

```python
counter = 0
while counter <= 10:
counter = counter + 1
print(counter)


  File "<ipython-input-1-d8461f52562c>", line 3
    counter = counter + 1
          ^
IndentationError: expected an indented block```

Let's use the while loop to create the table above:

In [51]:
C = -20                 # Initialise C
dC = 5                  # Increment for C within the loop
while C <= 40:          # Loop heading with condition
    F = (9.0/5)*C + 32  # 1st statement inside loop
    print(C, F)         # 2nd statement inside loop
    C = C + dC          # Increment C for the next iteration of the loop.

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


## Exercise 1.11: Make a Fahrenheit-Celsius conversion table
Write a program that uses a while loop to print out a table with Fahrenheit degrees 0, 10, 20, ..., 100 in the first column and the corresponding Celsius degrees in the second column.

In [52]:
# Uncomment and complete the code below. Do not change the names of variables.

Fahrenheit = 0
print ('Fahrenheit Celcius')
while Fahrenheit <= 100:
    Celsius = 5/9*(Fahrenheit-32)
    print ("%d %15.3f" %(Fahrenheit, Celsius))
    Fahrenheit = Fahrenheit + 10
    



Fahrenheit Celcius
0         -17.778
10         -12.222
20          -6.667
30          -1.111
40           4.444
50          10.000
60          15.556
70          21.111
80          26.667
90          32.222
100          37.778


In [53]:
ok.grade('question-1_11')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 1}

## Exercise 1.12: Write an approximate Fahrenheit-Celsius conversion table

Many people use an approximate formula for quickly converting Fahrenheit ($F$) to Celsius ($C$) degrees:</br></br>
$C \approx \hat{C} = \frac{F − 30}{2}$</br></br>
Modify the program from the previous exercise so that it prints three columns: $F$, $C$, and the approximate value $\hat{C}$.

In [54]:
# Uncomment and complete the code below. Do not change the names of variables.
Fahrenheit = 0
print ('Fahrenheit Celcius ApproxCelcius')
while Fahrenheit <= 100:
    Celsius = 5/9*(Fahrenheit-32)
    Celsius_approx = (Fahrenheit-30)/2
    print ('%5d %11.3f %5d'%(Fahrenheit, Celsius, Celsius_approx))
    Fahrenheit = Fahrenheit + 10
    

Fahrenheit Celcius ApproxCelcius
    0     -17.778   -15
   10     -12.222   -10
   20      -6.667    -5
   30      -1.111     0
   40       4.444     5
   50      10.000    10
   60      15.556    15
   70      21.111    20
   80      26.667    25
   90      32.222    30
  100      37.778    35


In [55]:
ok.grade('question-1_12')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Lists
So far, one variable has referred to one number (or string). Sometimes however we naturally have a collection of numbers, say
degrees −20, −15, −10, −5, 0, ..., 40. One way to store these values in a computer program would be to have one variable per value, i.e.

In [56]:
C1 = -20
C2 = -15
C3 = -10
# ...
C13 = 40

This is clearly a terrible solution, particularly if we have lots of values. A better way of doing this is to collect values together in a list:

In [57]:
C = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]

Now there is just one variable, **C**, holding all the values. Elements in a list are accessed via an index. List indices are always numbered as 0, 1, 2, and so forth up to the number of elements minus one:

In [58]:
mylist = [4, 6, -3.5]
print(mylist[0])
print(mylist[1])
print(mylist[2])
print(len(mylist))  # length of list

4
6
-3.5
3


Here are a few examples of operations that you can perform on lists:

In [59]:
C = [-10, -5, 0, 5, 10, 15, 20, 25, 30]
C.append(35) # add new element 35 at the end
print(C)

[-10, -5, 0, 5, 10, 15, 20, 25, 30, 35]


In [60]:
C=C+[40,45] # And another list to the end of C
print(C)

[-10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [61]:
C.insert(0, -15)     # Insert -15 as index 0
print(C)

[-15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [62]:
del C[2]             # delete 3rd element
print(C)

[-15, -10, 0, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [63]:
del C[2]             # delete what is now 3rd element
print(C)

[-15, -10, 5, 10, 15, 20, 25, 30, 35, 40, 45]


In [64]:
print(len(C))         # length of list

11


In [65]:
print(C.index(10))    # Find the index of the list with the value 10

3


In [66]:
print(10 in C)        # True only if the value 10 is stored in the list

True


In [67]:
print(C[-1])          # The last value in the list.

45


In [68]:
print(C[-2])          # The second last value in the list.

40


You can also extract sublists using ":"

In [69]:
print(C[5:])          # From index 5 to the end of the list.

[20, 25, 30, 35, 40, 45]


In [70]:
print(C[5:7])         # From index 5 up to, but not including index 7.

[20, 25]


In [71]:
print(C[7:-1])        # From index 7 up to the second last element.

[30, 35, 40]


In [72]:
print(C[:])           # [:] specifies the whole list.

[-15, -10, 5, 10, 15, 20, 25, 30, 35, 40, 45]


You can also unpack the elements of a list into seperate variables:

In [73]:
somelist = ['Curly', 'Larry', 'Moe']
stooge1, stooge2, stooge3 = somelist
print(stooge3, stooge2, stooge1)

Moe Larry Curly


## Exercise 1.13: Store odd numbers in a list

Step 1: Write a program that generates all odd numbers from 1 to *n*. For the purpose of testing, set *n=10* at the beginning of the program and use a while loop to compute the numbers. (Make sure that if *n* is an even number, the largest generated odd number is *n*-1.).

Step 2: Store the generated odd numbers in a list. Start with an empty list and use the same while loop where you generate each odd number, to append the new number to the list.

In [74]:
# Uncomment and complete code. Do not change the variable names.
n = 10
odd_list = []
k = 1
while k <=n:
    if k%2 != 0:
        odd_list.append(k)
    k = k + 1
print (odd_list)

[1, 3, 5, 7, 9]


In [75]:
ok.grade('question-1_13')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## For loops
We can visit each element in a list and process the element with some statements using a *for* loop, for example:

In [76]:
degrees = [0, 10, 20, 40, 100]
for C in degrees:
    print('list element:', C)
print('The degrees list has', len(degrees), 'elements')

list element: 0
list element: 10
list element: 20
list element: 40
list element: 100
The degrees list has 5 elements


Notice again how the statements to be executed within the loop must be indented! Let's now revisit the conversion table example using the *for* loop:

In [77]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
for C in Cdegrees:
    F = (9.0/5)*C + 32
    print(C, F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


We can easily beautify the table using the printf syntax that we encountered in the last lecture:

In [78]:
for C in Cdegrees:
    F = (9.0/5)*C + 32       
    print('%5d %5.1f' % (C, F))

  -20  -4.0
  -15   5.0
  -10  14.0
   -5  23.0
    0  32.0
    5  41.0
   10  50.0
   15  59.0
   20  68.0
   25  77.0
   30  86.0
   35  95.0
   40 104.0


It is also possible to rewrite the *for* loop as a *while* loop, i.e.,

```
for element in somelist:
           # process element
```

can always be transformed to a *while* loop
```
index = 0
while index < len(somelist):
    element = somelist[index]
    # process element
    index += 1
    ```

Taking the previous table example:

In [79]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
index = 0
while index < len(Cdegrees):
    C = Cdegrees[index]
    F = (9.0/5)*C + 32
    print('%5d %5.1f' % (C, F))
    index += 1

  -20  -4.0
  -15   5.0
  -10  14.0
   -5  23.0
    0  32.0
    5  41.0
   10  50.0
   15  59.0
   20  68.0
   25  77.0
   30  86.0
   35  95.0
   40 104.0


Rather than just printing out the Fahrenheit values, let's also store these computed values in a list of their own:

In [80]:
Cdegrees = [-20, -15, -10, -5, 0, 5, 10, 15, 20, 25, 30, 35, 40]
Fdegrees = []            # start with empty list
for C in Cdegrees:
    F = (9.0/5)*C + 32
    Fdegrees.append(F)   # add new element to Fdegrees
print(Fdegrees)

[-4.0, 5.0, 14.0, 23.0, 32.0, 41.0, 50.0, 59.0, 68.0, 77.0, 86.0, 95.0, 104.0]


In Python *for* loops usually loop over list values (elements), i.e.,

```
for element in somelist:
    ...process variable element
```

However, we can also loop over list indices:

```
for i in range(0, len(somelist), 1):
    element = somelist[i]
    ... process element or somelist[i] directly
    ```   

The statement *range(start, stop, inc)* generates a list of integers *start*, *start+inc*, *start+2\*inc*, and so on up to, but not including, *stop*. We can also write *range(stop)* as an abbreviation for *range(0, stop, 1)*:

In [81]:
print(range(3)) # same as range(0, 3, 1)

range(0, 3)


In [82]:
print(range(2, 8, 3))

range(2, 8, 3)


## List comprehensions
Consider this example where we compute two lists in a *for* loop:

In [83]:
n = 16
Cdegrees = [];  Fdegrees = []  # empty lists
for i in range(n):
    Cdegrees.append(-5 + i*0.5)
    Fdegrees.append((9.0/5)*Cdegrees[i] + 32)
print("Cdegrees = ", Cdegrees)
print("Fdegrees = ", Fdegrees)

Cdegrees =  [-5.0, -4.5, -4.0, -3.5, -3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5]
Fdegrees =  [23.0, 23.9, 24.8, 25.7, 26.6, 27.5, 28.4, 29.3, 30.2, 31.1, 32.0, 32.9, 33.8, 34.7, 35.6, 36.5]


As constructing lists is a very common requirement, the above way of doing it can become very tedious to both write and read. Therefore, Python has a compact construct, called list comprehension for generating lists from a *for* loop:

In [84]:
n = 16
Cdegrees = [-5 + i*0.5 for i in range(n)]
Fdegrees = [(9.0/5)*C + 32 for C in Cdegrees]
print("Cdegrees = ", Cdegrees)
print("Fdegrees = ", Fdegrees)

Cdegrees =  [-5.0, -4.5, -4.0, -3.5, -3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5]
Fdegrees =  [23.0, 23.9, 24.8, 25.7, 26.6, 27.5, 28.4, 29.3, 30.2, 31.1, 32.0, 32.9, 33.8, 34.7, 35.6, 36.5]


The general form of a list comprehension is:
```
somelist = [expression for element in somelist]
```

## Exercise 1.14: Create a list of even numbers ranging from 0 to 100 using a for loop.

In [85]:
# Use the variable name 'even_list' for testing purposes.
n = 102
even_list = []
for i in range(n):
    if i%2 == 0:
        even_list.append(i)
print("The list of even numbers is", even_list)
    
    


The list of even numbers is [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100]


In [86]:
ok.grade('question-1_14')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Exercise 1.15: Implement the sum function

The built-in Python function [sum](https://docs.python.org/3/library/functions.html#sum) takes a list as argument and computes the sum of the elements in the list:
```
sum([1,3,5,-5])

4
```
Implement your own version of sum.

In [87]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def my_sum(x):
    my_sum = 0
    for i in range (len(x)):
        my_sum +=x[i]
    return my_sum
print (my_sum([1,3,5,-5]))    

4


In [88]:
ok.grade('question-1_15')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 3}

## Exercise 1.16: Function that returns a list.

Write a function that creates a `list`, $t$ with `num` values ranging from `t_start` to `t_end` and returns the `list` of $y$ values calculated using the formula:
$$y(t) = v_0t − gt^2.$$
Specify the keyword arguments $v0=6.0$ and $g=9.81$.

In [89]:
# Uncomment and complete this code - keep the names the same for testing purposes. 
def distance(t_start, t_end, num, v0=6.0, g=9.81):
    d = (t_end - t_start)/(num-1)
    t = [t_start]
    term = t_start
    y = []
    n = 1
    i = 0
    while n < num:
        term = t_start + n*d
        t.append(term)
        n+=1
    for i in range(len(t)):
        y.append(v0*t[i] - g * (t[i]**2))
        i+=1
    return y
print(distance(0.1,0.6,6))    
        
        

[0.5019, 0.8076000000000001, 0.9171, 0.8304, 0.5474999999999999, 0.06839999999999957]


In [90]:
ok.grade('question-1_16')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
question 1.16 > Suite 2 > Case 2

import numpy as np
>>> np.allclose(distance(1, 6, 5, 3),[-6.8100000000000005, -8.570100000000002, -10.5264, -12.6789, -15.0276])
False

# Error: expected
#     True
# but got
#     False

Run only this test case with "python3 ok -q question-1_16 --suite 2 --case 2"
---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 1
[ooooook....] 66.7% passed



{'failed': 1, 'locked': 0, 'passed': 2}

## Exercise 1.17: Cumulative sum

Write a function that returns the cumulative sum of the numbers in a list. The function should return a list, whose `i`th element is the sum of the input list up to and including the `i`th element.

For example, for the list `[1, 4, 2, 5, 3]` should return `[1, 5, 7, 12, 15]`.

In [91]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def my_cumsum(l):
    my_cumsum = []
    b = 0
    for i in range (len(l)):
        b +=l[i]
        my_cumsum.append(b)
    return my_cumsum
print(my_cumsum([1, 4, 2, 5, 3]))


[1, 5, 7, 12, 15]


In [92]:
ok.grade('question-1_17')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Exercise 1.18: Bouncing ball

A rubber ball is dropped from a height `h_0`. After each bounce, the height it rebounds to decreases by 10%; i.e. after one bounce it reaches `0.9*h_0`, after two bounces it reaches `0.9*0.9*h_0`, etc.

Write a function that returns a list of the maximum heights of the ball after each bounce (including after 0 bounces, i.e. its initial height), until either the ball has bounced `n` times *or* its maximum height falls below `h_1`. The function should take `h_0`, `h_1` and `n` as keyword arguments, with default values of 1.0, 0.3 and 10 respectively.

In [93]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def compute_heights(h_0=1.0, h_1=0.3, n=10):
    b = 1
    a = 0
    c = []
    d = 1.0
    while a <= n and d>h_1:
        d = b*h_0
        c.append(d)
        b=b*0.9
        a=a+1
    return c
print(compute_heights())

[1.0, 0.9, 0.81, 0.7290000000000001, 0.6561000000000001, 0.5904900000000002, 0.5314410000000002, 0.47829690000000014, 0.43046721000000016, 0.38742048900000015, 0.34867844010000015]


In [94]:
compute_heights(h_0=0.5)

[0.5,
 0.45,
 0.405,
 0.36450000000000005,
 0.32805000000000006,
 0.2952450000000001]

In [95]:
ok.grade('question-1_18')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 4}

## Exercise 1.19: Calculate Pi

A formula for $\pi$ is given by the Gregory-Leibniz series:

$$\pi = 4 \left(\frac{1}{1} - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} + ...  \right)$$

Note that the denominators of the terms in this series are the positive odd numbers. Follow the guidelines below to calculate $\pi$; each of the first three steps can be completed using a single list comprehension.

Step 1:

Modify your answer to Exercise 1.14 to produce a list of the first `n` odd numbers, for `n=100`.

Step 2:

Make a list of the signs of each term, i.e. `[1, -1, 1, -1, ...]`. Store the result in a list named `signs`. Hint: think about the value of $(-1)^i$ for integer $i$.

Step 3:

Using the results of steps 1 and 2, make a list of the first `n` terms in the above series, and store in a variable named `series_terms`.

Step 4:

Use your `my_sum` function from the previous exercise to sum this series, and multiply by 4. Store this value in a variable named `my_pi`.

In [96]:
# Uncomment and complete this code - keep the names the same for testing purposes. 
n = 100000
signs=[]
odd_integers = []
series_terms = []
my_sum = 0
b = 0
for i in range(n):
    if i%2 != 0:
        odd_integers.append(i)
#print(odd_integers)
a = 0
while a<(n/2):
    signs.append((-1)**a)
    a += 1
#print (signs)
for b in range(len(signs)):
    series_terms.append(signs[b]/odd_integers[b])
    b +=1
#print(series_terms)
for i in range(len(series_terms)):
    my_sum += series_terms[i]
my_pi = 4 * my_sum   
print(my_pi)

    


#signs = [...]

#series_terms = [...]

#my_pi = ...

#print(my_pi)

3.1415726535897814


In [97]:
ok.grade('question-1_19')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 1}

#### ok.score()